lab_4_retrieval_w_clustering/main.py

```diff
@@ -182,14 +257,31 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if not (isinstance(query, str) and isinstance(n_neighbours, int) and
+                n_neighbours > 0):
+            raise ValueError
+        query_vector = self._index_document(query)
```

Comment:
why do you use `_index_document`? Why not tokenize?

Reply:
To tokenize, we have to address the _tokenizer, _index_document cam be used from the search engine itself. But I changed the method anyway, as they seem to work the same.